File: rust/2021/src/bin/day_07.rs
use std::fs;

fn min_fuel(crabs: &[i32], constant: bool) -> i32 {
    let (lower, upper) = (crabs[0], crabs[crabs.len() - 1]);
    (lower..=upper).map(|pos| {
        crabs.iter().map(|&x| {
            let dist = (x - pos).abs();
            if constant { dist } else { dist * (dist + 1) / 2}
        }).sum()
    }).min().unwrap()
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_07.txt").unwrap();
    let mut crabs: Vec<i32> = input.trim().split(',').map(|s| s.parse().unwrap()).collect();
    crabs.sort();
    println!("Part 1: {}", min_fuel(&crabs, true));
    println!("Part 2: {}", min_fuel(&crabs, false));
}

File: haskell/2021/day_07.hs
import qualified Data.List.Split as Split

minFuel :: Bool -> [Int] -> Int
minFuel constant crabs = minimum $ (\pos ->
        sum $ (\x -> let dist = abs (x - pos)
            in if constant then dist else dist * (dist + 1) `div` 2
        ) <$> crabs
    ) <$> [minimum crabs .. maximum crabs]

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_07.txt"
    let crabs = read <$> Split.splitOn "," (head $ words text)
    putStrLn $ "Part 1: " ++ show (minFuel True crabs)
    putStrLn $ "Part 2: " ++ show (minFuel False crabs)
